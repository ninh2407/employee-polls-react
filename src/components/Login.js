import { connect } from 'react-redux';
import { setAuthedUser } from '../actions/authedUser';
import { useNavigate } from "react-router-dom";

const Login = (props) => {
    const navigate = useNavigate();

    const handleLogin = (e) => {
        e.preventDefault();
        props.dispatch(setAuthedUser(document.getElementById("userId").value));
        navigate("/homepage");
    } 
    return (
        <div className='container'>
            <div> Please select user </div>
            <select id="userId">
                {Object.keys(props.listUsers).map((id) => {
                    return <option key={id} value={id}>{id}</option>
                })}
            </select>
            <button onClick={handleLogin}>Login</button>
        </div>
    );
};

const mapStateToProps = ({ listUsers, authedUser }) => ({
    listUsers,
    authedUser
});

export default connect(mapStateToProps)(Login);