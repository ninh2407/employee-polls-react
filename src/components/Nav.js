import { connect } from 'react-redux';
import { Link } from "react-router-dom";
import { setAuthedUser } from '../actions/authedUser';

const Nav = (props) => {
    const logOut = () => {
        props.dispatch(setAuthedUser(""));
    }

    return (
        <nav className="nav">
            <ul>
                <li>
                    <Link to="/homepage">Home Page</Link>
                </li>
                <li>
                    <Link to="/leaderboard">Leader Board</Link>
                </li>
                <li>
                    <Link to="/add">New</Link>
                </li>
                <li>
                    <Link to="/" onClick={logOut}>{props.authedUser ? "Log out" : "Log in"}</Link>
                </li>
                {props.authedUser && (
                <li>
                    User : {props.authedUser}
                </li>
            )}
            </ul>
        </nav>
    );
};

const mapStateToProps = ({ authedUser }) => ({
    authedUser
});

export default connect(mapStateToProps)(Nav);