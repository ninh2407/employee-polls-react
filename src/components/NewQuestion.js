import { connect } from 'react-redux';
import { useNavigate } from "react-router-dom";
import { handleAddQuestion } from '../actions/questions';
import { handleInitData } from '../actions/init';

const NewQuestion = (props) => {
    const navigate = useNavigate();

    const handleSaveQuestion = (e) => {
        e.preventDefault();
        let question = { 
            "optionOneText": document.getElementById("firstOption").value,
            "optionTwoText": document.getElementById("secondOption").value, 
            "author" : props.authedUser
        }
        console.log(question);
        props.dispatch(handleAddQuestion(question));
        props.dispatch(handleInitData());
        navigate("/homepage");
    }
    return (
        <div className='container'>
            <h2>
                Would You Rather
            </h2>
            <h4>
                Create Your Own Poll
            </h4>
            <div>
                <div>
                    First Option
                </div>
                <div>
                    <input type="text" data-testid="firstOption" id="firstOption" />
                </div>
            </div>
            <div>
                <div>
                    Second Option
                </div>
                <div>
                    <input type="text" data-testid="secondOption" id="secondOption" />
                </div>
            </div>
            <button onClick={handleSaveQuestion}>Submit</button>
        </div>
    );
};

const mapStateToProps = ({ listUsers, authedUser }) => ({
    listUsers,
    authedUser
});

export default connect(mapStateToProps)(NewQuestion);