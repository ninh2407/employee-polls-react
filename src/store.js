import { configureStore } from '@reduxjs/toolkit';
import authedUser from "./reducers/authedUser";
import listUsers from "./reducers/users";
import questions from "./reducers/questions";

export const store = configureStore({
    reducer: {
        listUsers,
        authedUser,
        questions,
    },
});