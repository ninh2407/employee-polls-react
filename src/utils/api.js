import {
  _getUsers,
  _getQuestions,
  _saveQuestion,
  _saveQuestionAnswer
} from "./_DATA";

export function getInitData() {
  return Promise.all([
    _getUsers(),
    _getQuestions(),
  ]).then(([users, questions]) => ({
    users,
    questions,
  }))
}

export function getQuestions() {
  return _getQuestions();
}

export function saveQuestion(info) {
  return _saveQuestion(info);
}

export function saveQuestionAnswer(info) {
  return _saveQuestionAnswer(info);
}