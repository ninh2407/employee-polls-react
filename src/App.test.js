import { _saveQuestion, _saveQuestionAnswer } from './utils/_DATA';
import { fireEvent, render } from '@testing-library/react';
import Nav from './components/Nav';
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from './store';
import NewQuestion from './components/NewQuestion';
import Login from './components/Login';

describe('_saveQuestion', () => {

  it('the saved question is returned and all expected fields are populated when correctly formatted data is passed to the function', async () => {
    var question = {
      "optionOneText": "optionOneText",
      "optionTwoText": "optionTwoText",
      "author": "johndoe"
    }
    var result = await _saveQuestion(question);
    expect(result.author).toEqual(question.author);
    expect(result.optionOne.text).toEqual(question.optionOneText);
    expect(result.optionTwo.text).toEqual(question.optionTwoText);
  })

  it('the saved question is returned and all expected fields are populated when correctly formatted data is passed to the function', async () => {
    var question = {
      "optionOneText": "banana",
      "optionTwoText": "apple",
      "author": "tylermcginnis"
    }
    var result = await _saveQuestion(question);
    expect(result.author).toEqual(question.author);
    expect(result.optionOne.text).toEqual(question.optionOneText);
    expect(result.optionTwo.text).toEqual(question.optionTwoText);
  })

  it('the saved question is returned and all expected fields are populated when correctly formatted data is passed to the function', async () => {
    var question = {
      "optionOneText": "facebook",
      "optionTwoText": "top top",
      "author": "sarahedo"
    }
    var result = await _saveQuestion(question);
    expect(result.author).toEqual(question.author);
    expect(result.optionOne.text).toEqual(question.optionOneText);
    expect(result.optionTwo.text).toEqual(question.optionTwoText);
  })

  it('an error is returned if incorrect data is passed to the function', async () => {
    var question = {
      "optionOneText": "optionOneText",
      "optionTwoText": "optionTwoText",
      "author": "unknow"
    }
    await expect(_saveQuestion(question)).rejects.toEqual("user id not found");
  })

})

describe('_saveQuestionAnswer', () => {

  it('true is returned when correctly formatted data is passed to the function', async () => {
    var answer = {
      "authedUser": "johndoe",
      "qid": "8xf0y6ziyjabvozdd253nd",
      "answer": "optionTwo"
    }

    await expect(_saveQuestionAnswer(answer)).toBeTruthy();
  })

  it('true is returned when correctly formatted data is passed to the function', async () => {
    var answer = {
      "authedUser": "tylermcginnis",
      "qid": "8xf0y6ziyjabvozdd253nd",
      "answer": "optionOne"
    }

    await expect(_saveQuestionAnswer(answer)).toBeTruthy();
  })

  it('an error is returned if incorrect data is passed to the function', async () => {
    var answer = {
      "authedUser": "unknow",
      "qid": "8xf0y6ziyjabvozdd253nd",
      "answer": "optionTwo"
    }
    await expect(_saveQuestionAnswer(answer)).rejects.toEqual("user id not found");
  })


  describe('snapshot', () => {

    it('Snapshot nav', () => {
      var component = render(
        <Provider store={store}>
          <BrowserRouter>
            <Nav />
          </BrowserRouter>
        </Provider>
      );
      expect(component).toMatchSnapshot();
    })

    it('Snapshot login page', () => {
      var component = render(
        <Provider store={store}>
          <BrowserRouter>
            <Login />
          </BrowserRouter>
        </Provider>
      );
      expect(component).toMatchSnapshot();
    })

  })

  describe('firechange new question', () => {

    it('firechange new question', () => {
      var component = render(
        <Provider store={store}>
          <BrowserRouter>
            <NewQuestion />
          </BrowserRouter>
        </Provider>
      );

      var inputOptionOne = component.getByTestId("firstOption");
      var inputOptionTwo = component.getByTestId("secondOption");

      fireEvent.change(inputOptionOne, { target: { value: 'batman' } });
      fireEvent.change(inputOptionTwo, { target: { value: 'superman' } });
      expect(inputOptionOne.value).toBe("batman");
      expect(inputOptionTwo.value).toBe("superman");
    })
  })

})
