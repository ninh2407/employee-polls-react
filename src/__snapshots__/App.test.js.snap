// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_saveQuestionAnswer snapshot Snapshot login page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div>
           Please select user 
        </div>
        <select
          id="userId"
        />
        <button>
          Login
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div>
         Please select user 
      </div>
      <select
        id="userId"
      />
      <button>
        Login
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`_saveQuestionAnswer snapshot Snapshot nav 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="nav"
      >
        <ul>
          <li>
            <a
              href="/homepage"
            >
              Home Page
            </a>
          </li>
          <li>
            <a
              href="/leaderboard"
            >
              Leader Board
            </a>
          </li>
          <li>
            <a
              href="/add"
            >
              New
            </a>
          </li>
          <li>
            <a
              href="/"
            >
              Log in
            </a>
          </li>
          
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="nav"
    >
      <ul>
        <li>
          <a
            href="/homepage"
          >
            Home Page
          </a>
        </li>
        <li>
          <a
            href="/leaderboard"
          >
            Leader Board
          </a>
        </li>
        <li>
          <a
            href="/add"
          >
            New
          </a>
        </li>
        <li>
          <a
            href="/"
          >
            Log in
          </a>
        </li>
        
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
